### PR TITLE
Dump commands in internal code using command printing functions.

### PR DESCRIPTION
**Pull request by abdoo8080**
This is work towards migrating commands to the new API. Internal code that creates command objects just for dumping is replaced with direct calls to functions that print the those commands.

### CODE REVIEW

File: src/theory/bv/abstraction.cpp [hunk 1]
```diff
@@ -694,10 +696,14 @@ Node AbstractionModule::simplifyConflict(TNode conflict) {
   if (Dump.isOn("bv-abstraction")) {
     NodeNodeMap seen;
     Node c = reverseAbstraction(conflict, seen);
-    Dump("bv-abstraction") << PushCommand();
-    Dump("bv-abstraction") << AssertCommand(c.toExpr());
-    Dump("bv-abstraction") << CheckSatCommand();
-    Dump("bv-abstraction") << PopCommand();
+
+    Printer* printer = smt::currentSmtEngine()->getPrinter();
```
Comment (ajreynol) [inline src/theory/bv/abstraction.cpp:700]: You can delete this block (everything related to the trace "bv-abstraction").

File: src/theory/bv/abstraction.cpp [hunk 2]
```diff
@@ -746,10 +752,14 @@ Node AbstractionModule::simplifyConflict(TNode conflict) {
 
     NodeNodeMap seen;
     Node nc = reverseAbstraction(new_conflict, seen);
-    Dump("bv-abstraction") << PushCommand();
-    Dump("bv-abstraction") << AssertCommand(nc.toExpr());
-    Dump("bv-abstraction") << CheckSatCommand();
-    Dump("bv-abstraction") << PopCommand();
+
+    Printer* printer = smt::currentSmtEngine()->getPrinter();
```
Comment (ajreynol) [inline src/theory/bv/abstraction.cpp:756]: Ditto, delete.

File: src/theory/bv/abstraction.cpp [hunk 3]
```diff
@@ -840,10 +850,14 @@ void AbstractionModule::generalizeConflict(TNode conflict, std::vector<Node>& le
       if (Dump.isOn("bv-abstraction")) {
         NodeNodeMap seen;
         Node l = reverseAbstraction(lemma, seen);
-        Dump("bv-abstraction") << PushCommand();
-        Dump("bv-abstraction") << AssertCommand(l.toExpr());
-        Dump("bv-abstraction") << CheckSatCommand();
-        Dump("bv-abstraction") << PopCommand();
+
+        Printer* printer = smt::currentSmtEngine()->getPrinter();
```
Comment (ajreynol) [inline src/theory/bv/abstraction.cpp:854]: Ditto, delete.

File: src/theory/bv/bv_subtheory_algebraic.cpp [hunk 4]
```diff
@@ -324,11 +328,16 @@ bool AlgebraicSolver::check(Theory::Effort e)
     if (Dump.isOn("bv-algebraic")) {
       Node expl = d_explanations[id];
       Node query = utils::mkNot(nm->mkNode(kind::IMPLIES, expl, fact));
-      Dump("bv-algebraic") << EchoCommand("ThoeryBV::AlgebraicSolver::substitution explanation");
-      Dump("bv-algebraic") << PushCommand();
-      Dump("bv-algebraic") << AssertCommand(query.toExpr());
-      Dump("bv-algebraic") << CheckSatCommand();
-      Dump("bv-algebraic") << PopCommand();
+
+      Printer* printer = smt::currentSmtEngine()->getPrinter();
```
Comment (ajreynol) [inline src/theory/bv/bv_subtheory_algebraic.cpp:332]: You can delete everything related to the dump "bv-algebraic".

**Review by ajreynol — CHANGES_REQUESTED** [review]
A few comments.

Also, we should minimize the usage of ` smt::currentSmtEngine()`, since it relies on a global pointer and will lead to design issues if and when we change our policy on SmtEngine scoping. We should have an object to pass around that maintains the printer/output stream.  The object would be owned by SmtEngine and then passed to all modules that require dumping commands.

File: src/preprocessing/preprocessing_pass.cpp [hunk 5]
```diff
@@ -39,8 +40,14 @@ void PreprocessingPass::dumpAssertions(const char* key,
   if (Dump.isOn("assertions") && Dump.isOn(std::string("assertions:") + key))
   {
     // Push the simplified assertions to the dump output stream
-    for (const auto& n : assertionList) {
-      Dump("assertions") << AssertCommand(Expr(n.toExpr()));
+    const Printer& printer =
+        d_preprocContext->getSmt()->getOutputManager().getPrinter();
```
Comment (ajreynol) [inline src/preprocessing/preprocessing_pass.cpp:44]: Nit: could cache `OutputManager& om = d_preprocContext->getSmt()->getOutputManager()`.

File: src/smt/dump.cpp [hunk 6]
```diff
@@ -248,12 +260,6 @@ t-explanations [non-stateful]\n\
 bv-rewrites [non-stateful]\n\
 + Output correctness queries for all bitvector rewrites\n\
 \n\
-bv-abstraction [non-stateful]\n\
```
Comment (ajreynol) [inline src/smt/dump.cpp:251]: Should also remove documentation for the other removed ones, e.g. `state`, not sure if there are others.

File: src/smt/preprocessor.cpp [hunk 7]
```diff
@@ -128,7 +129,8 @@ Node Preprocessor::simplify(const Node& node, bool removeItes)
   Trace("smt") << "SMT simplify(" << node << ")" << endl;
   if (Dump.isOn("benchmark"))
   {
-    Dump("benchmark") << SimplifyCommand(node.toExpr());
+    d_smt.getOutputManager().getPrinter().toStreamCmdSimplify(
+        *d_smt.getOptions().getOut(), node);
```
Comment (ajreynol) [inline src/smt/preprocessor.cpp:133]:
```suggestion
        *d_smt.getOutputManager().getDumpOut(), node);
```

Comment (ajreynol) [reply src/smt/preprocessor.cpp:133]: Also can make this more compact by caching `OutputManager& om = d_smt.getOutputManager()`.

File: src/smt/process_assertions.cpp [hunk 8]
```diff
@@ -562,7 +564,8 @@ void ProcessAssertions::dumpAssertions(const char* key,
     for (unsigned i = 0; i < assertionList.size(); ++i)
     {
       TNode n = assertionList[i];
-      Dump("assertions") << AssertCommand(Expr(n.toExpr()));
+      d_smt.getOutputManager().getPrinter().toStreamCmdAssert(
+          *d_smt.getOptions().getOut(), n);
```
Comment (ajreynol) [inline src/smt/process_assertions.cpp:568]: Ditto, should just use output manager dump output for consistency.

File: src/smt/smt_engine.cpp [hunk 9]
```diff
@@ -285,12 +297,13 @@ void SmtEngine::finishInit()
...
+      getOutputManager().getPrinter().toStreamCmdComment(
+          *getOptions().getOut(),
           "CVC4 always dumps the most general, all-supported logic (below), as "
           "some internals might require the use of a logic more general than "
-          "the input.")
-                        << SetBenchmarkLogicCommand(
-                               everything.getLogicString());
+          "the input.");
+      getOutputManager().getPrinter().toStreamCmdSetBenchmarkLogic(
+          *getOptions().getOut(), everything.getLogicString());
```
Comment (ajreynol) [inline src/smt/smt_engine.cpp:306]: Ditto, use output manager dump out.

File: src/smt/smt_engine.cpp [hunk 10]
```diff
@@ -404,8 +417,8 @@ void SmtEngine::setLogic(const std::string& s)
     // dump out a set-logic command
     if (Dump.isOn("raw-benchmark"))
     {
-      Dump("raw-benchmark")
-          << SetBenchmarkLogicCommand(d_logic.getLogicString());
+      getOutputManager().getPrinter().toStreamCmdSetBenchmarkLogic(
+          *getOptions().getOut(), d_logic.getLogicString());
```
Comment (ajreynol) [inline src/smt/smt_engine.cpp:421]: Ditto.

File: src/smt/smt_engine.cpp [hunk 11]
```diff
@@ -459,12 +472,14 @@ void SmtEngine::setInfo(const std::string& key, const CVC4::SExpr& value)
...
       string s = value.getValue();
-      BenchmarkStatus status =
-        (s == "sat") ? SMT_SATISFIABLE :
-          ((s == "unsat") ? SMT_UNSATISFIABLE : SMT_UNKNOWN);
-      Dump("benchmark") << SetBenchmarkStatusCommand(status);
+      Result::Sat status =
+          (s == "sat") ? Result::SAT
+                       : ((s == "unsat") ? Result::UNSAT : Result::SAT_UNKNOWN);
+      getOutputManager().getPrinter().toStreamCmdSetBenchmarkStatus(
+          *getOptions().getOut(), status);
```
Comment (ajreynol) [inline src/smt/smt_engine.cpp:479]: Ditto

File: src/smt/smt_engine.cpp [hunk 12]
```diff
@@ -459,12 +472,14 @@ void SmtEngine::setInfo(const std::string& key, const CVC4::SExpr& value)
...
     } else {
-      Dump("benchmark") << SetInfoCommand(key, value);
+      getOutputManager().getPrinter().toStreamCmdSetInfo(
+          *getOptions().getOut(), key, value);
```
Comment (ajreynol) [inline src/smt/smt_engine.cpp:482]: Ditto

File: src/smt/smt_engine.cpp [hunk 13]
```diff
@@ -774,16 +789,15 @@ void SmtEngine::defineFunctionsRec(
...
-      tFormals.emplace_back(api::exprVectorToTerms(d_solver, formal));
+      nFormals.emplace_back(exprVectorToNodes(formal));
     }
-    std::vector<api::Term> tFormulas =
-        api::exprVectorToTerms(d_solver, formulas);
-    Dump("raw-benchmark") << DefineFunctionRecCommand(
-        d_solver, tFuncs, tFormals, tFormulas, global);
+    std::vector<Node> nFormulas = exprVectorToNodes(formulas);
+    getOutputManager().getPrinter().toStreamCmdDefineFunctionRec(
+        *getOptions().getOut(), nFuncs, nFormals, nFormulas);
```
Comment (ajreynol) [inline src/smt/smt_engine.cpp:800]: Ditto

File: src/smt/smt_engine.cpp [hunk 14]
```diff
@@ -930,7 +944,11 @@ void SmtEngine::notifyPostSolvePost()
 
 Result SmtEngine::checkSat(const Expr& assumption, bool inUnsatCore)
 {
-  Dump("benchmark") << CheckSatCommand(assumption);
+  if (Dump.isOn("benchmark"))
+  {
+    getOutputManager().getPrinter().toStreamCmdCheckSat(*getOptions().getOut(),
```
Comment (ajreynol) [inline src/smt/smt_engine.cpp:949]: Ditto

**Review by ajreynol — CHANGES_REQUESTED** [review]
The changes involving passing output manager and its definition look good.  My main request is to make sure we use output streams consistently, currently, some places this PR uses `getOptions().getOut()` and other places it uses `getOuputManager().getDumpOut()`. We should always use the latter for dumping commands.

File: src/smt/dump.cpp (hunk 6 above)
Comment (abdoo8080) [reply src/smt/dump.cpp:251]: Discussed offline.

**Review by ajreynol — APPROVED** [review]
LGTM, thanks for the updates.

**Pull request merged**